src/python/grpcio/grpc/_observability.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2023 The gRPC authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import abc
+import logging
+import threading
+from typing import Any, Generic, Optional, TypeVar
+
+from grpc._cython import cygrpc as _cygrpc
+
+_LOGGER = logging.getLogger(__name__)
+
+_channel = Any  # _channel.py imports this module.
+ClientCallTracerCapsule = TypeVar('ClientCallTracerCapsule')
+ServerCallTracerFactoryCapsule = TypeVar('ServerCallTracerFactoryCapsule')
+
+_lock: threading.RLock = threading.RLock()
+_grpc_observability_stub: Optional[ObservabilityPlugin] = None  # pylint: disable=used-before-assignment
+
+
+class ObservabilityPlugin(Generic[ClientCallTracerCapsule,
+                                  ServerCallTracerFactoryCapsule],
+                          metaclass=abc.ABCMeta):
+    """
+    Note: Any future methods added to this interface cannot have the @abc.abstractmethod annotation.
+    """
+    _tracing_enabled: bool = False
+    _stats_enabled: bool = False
+
+    @abc.abstractmethod
+    def create_client_call_tracer(
+            self, method_name: bytes) -> ClientCallTracerCapsule:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def delete_client_call_tracer(
+            self, client_call_tracer: ClientCallTracerCapsule) -> None:
+        # delte client call tracer have to be called on o11y package side.
+        # Call it for both segregated and integrated call (`_process_integrated_call_tag`)
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def save_trace_context(self, trace_id: str, span_id: str,
+                           is_sampled: bool) -> None:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def create_server_call_tracer_factory(
+            self) -> ServerCallTracerFactoryCapsule:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def record_rpc_latency(self, method: str, rpc_latency: float,
+                           status_code: Any) -> None:
+        raise NotImplementedError()
+
+    def enable_tracing(self, enable: bool) -> None:
```

Comment:
The name `enable_tracing` implies that this unconditionally enables tracing and therefore takes no arguments. Instead, this should probably be named `set_tracing`. Same for `enable_stats` below.